Overhaul the API

bugaevc opened the pull request:
Overhaul the API as requested in #33. In particular,
* Remove `MmapView` and `MmapViewSync` types
* Define two types - `Mmap` and `MmapMut` for readable and readable/writable mappings respectively. They implement `Deref<Target = [u8]>`/`DerefMut` as appropriate.
    * `Mmap` has no `flush` methods.
    * Alongside "raw" `set_protection()`, provide `make_mut()` and `make_read_only()` to safely convert between `Mmap` and `MmapMut`.
* Get rid of multiple `open`, `open_path`, `anonymous_with_options`, ... methods. Instead, provide two `OpenOptions`-style builders, one for file-backed and one for anonymous mappings.
    * Remove support for opening files by their path. Always use `std::fs::File`.
    * On Unix, `memmap::file()` is `unsafe`. This seems to be the best approach, although it is always safe to use it with `ReadCopy`.
    * To improve ergonomics, automatically pick a suitable `Protection` unless specified explicitly.
        * This makes it ergonomic enough so there's no need for additional convenience constructors.
* Convert the tests and the cat example to use the new API.
    * `use super as memmap` doesn't seem to work, rustc says that `super` is not found because it's the root module, WTF? Yet, `use super::*` works. Check out the workaround I came up with.

bugaevc commented:
The build is failing on Windows because of unnecessary `unsafe`s (should I just `#[allow]` them?) and on Rust 1.8 because I used newer features such as the `?` operator.

jonas-schievink commented:
What happens if I create a `MmapMut` and call `set_protection(Protection::Read)`? Will a write just segfault?

bugaevc commented:
> What happens if I create a MmapMut and call set_protection(Protection::Read)? Will a write just segfault?

No, `set_protection()` will [return `Err`](https://github.com/bugaevc/memmap-rs/blob/af52ac9d9395bb3971a105b0e151c005904f4e2d/src/lib.rs#L448-L451) appropriately.

An alternative approach would be to split `Protection` into separate enums for mutable and immutable options. That would allow for things like this to be enforced statically, but it seems to make APIs less ergonomic.

jonas-schievink commented:
Ah, no, that's fine. I didn't see it in the diff because it wasn't getting rendered.

danburkert commented on src/lib.rs:
```diff
@@ -47,115 +45,342 @@ pub enum Protection {
+
+    /// Actually map this anonymous mapping into the address space.
+    ///
+    /// This methos returns an immutable mapping, see [`map_mut()`](#method.map_mut)
```
typo: method

danburkert commented on src/lib.rs:
```diff
@@ -47,115 +45,342 @@ pub enum Protection {
+    ///
+    /// This method returns `Err` when the underlying system call fails, which can happen for
+    /// a variety of reasons, such as when you don't have the necessary permissions for the file.
+    pub fn map(&self) -> Result<Mmap> {
```
Does this constructor actually make sense?  What's the usecase for creating a read-only anonymous memory map?

bugaevc replied:
Yeah, I thought about this too, but decided to leave it here for consistency. I should however document that this is not what you want most of the time.

danburkert replied:
I think we should go ahead and remove it unless there's a compelling reason to keep it at this point.  We can always add it back later if a case can be made.

danburkert commented on src/lib.rs:
```diff
@@ -47,115 +45,342 @@ pub enum Protection {
+/// [`Protection::ReadCopy`](enum.Protection.html#variant.ReadCopy)
+/// is always safe.
+#[cfg(unix)]
+pub unsafe fn file(file: &File) -> FileMappingOptions {
```
I would strongly prefer if this API were the same across all platforms, even if that means marking it unsafe on Windows.

bugaevc replied:
Okay.

danburkert commented on src/lib.rs:
```diff
@@ -47,115 +45,342 @@ pub enum Protection {
+/// In particular, it is **undefined behavior** in Rust for the memory to be
+/// modified by some other code while there's a reference to it.
+///
+/// Note, however, that using this with
```
This has never really been clear to me.  The Linux `mmap` manpage says:

> It is unspecified whether changes made to the file after the mmap() call are visible in the mapped region.

That doesn't seem like it will play well with Rust's safety semantics.

bugaevc replied:
Ah, right. I'll change the message to say it's unsafe even with cow.

danburkert commented on src/lib.rs:
```diff
@@ -207,395 +432,112 @@ impl Mmap {
+    ///
+    /// This method *also* returns `Err` with `ErrorKind` set to `InvalidInput` if the specified
+    /// protection does not allow the mapping to be mutable.
+    pub fn set_protection(&mut self, protection: Protection) -> Result<()> {
```
So this is just for switching from ReadWrite to ReadCopy and vice versa?  That seems kinda strange - we should probably document what that even means.  I'm surprised the OS allows it.

danburkert commented on src/lib.rs:
```diff
@@ -207,395 +432,112 @@ impl Mmap {
-            len: self.len,
-        }
+    /// This method will **not** return `Err` if the passed `protection` is mutable.
+    pub fn make_read_only(mut self, protection: Protection) -> Result<Mmap> {
```
Perhaps this could be `make_read_only`, and `make_read_execute`?  Are there any other valid outcomes from this?

danburkert commented on src/lib.rs:
```diff
@@ -207,395 +432,112 @@ impl Mmap {
+    #![allow(unused_unsafe)]
+
+    mod memmap {
+        pub use super::super::*;
```
Is it not possible to do a `use super::*;` here without the extra mod?  I've never seen anything like this.

bugaevc replied:
It is, I just wanted the tests code to say `memmap::file` instead of just `file`, etc. As I said above, `use super as memmap` doesn't work, this is either a bug or me misunderstanding how the module system works.

danburkert commented on src/windows.rs:
```diff
@@ -78,13 +77,7 @@ impl MmapInner {
 
-    /// Check if a file needs to be writable for copy-on-write mode.
-    #[inline]
-    pub fn needs_write_for_copy() -> bool {
```
Bad rebase?

bugaevc replied:
No, actually, this method isn't used anymore, because we require for the file to be opened in advance.

danburkert replied:
ah, I see

danburkert commented on src/lib.rs:
```diff
@@ -47,115 +45,342 @@ pub enum Protection {
+                ErrorKind::InvalidInput,
+                "Invalid protection for a mutable mapping",
+            )),
+            Protection::ReadWrite | Protection::ReadCopy => Ok(
```
Does ReadCopy even make sense for an Anonymous mapping?  I'm having a hard time reasoning through some of these semantics.

bugaevc replied:
No, not really; again, it's here for consistency. Do you think we should work out what combinations are actually useful and only allow those?

danburkert replied:
Yah, I think eventually we should, but we can punt for now and do it in a follow-up commit.

bugaevc commented:
We could get rid of protection in public API entirely, instead exposing a bunch of methods: {map, make}_read_{only, write, execute, copy_on_write} for initial mapping and conversions *where they make sense*.

bugaevc commented:
Rebased to separate `MmapView` removal and addressed a couple of pointed out issues: naming (`Mapping` -> `Mmap` in `Options`), a typo ("methos") and API difference between Unix and Windows (it's now always `unsafe` to mmap a file).

Please let me know what you think about my suggestion above.

danburkert commented:
I merged the view removal commit, so you may need to rebase over that.

bugaevc commented:
* Removed `AnonymousMmapOptions::map()` and the corresponding test.
* Made it compile under 1.8.0 (why do you need that?)

danburkert commented:
Awesome, thanks.   As for Rust 1.8, there is no hard and fast requirement except that that's the status quo, so if it's not too onerous to maintain I'd prefer to.   This crate is likely to be a widely used dependency, and if it forces a certain rustic version, then it's forced on everyone who uses it as well.   Basically the same reason that libc maintains long backwards compat windows.

Will take a look and likely merge tonight or over the weekend.   Thanks again for the help!

bugaevc commented:
Great! So, if you think this:
> We could get rid of protection in public API entirely, instead exposing a bunch of methods: `{map, make}_read_{only, write, execute, copy_on_write}` for initial mapping and conversions *where they make sense.*

is worth a `try!` 😄 , I'll make a separate PR, is that OK?

bugaevc commented:
Also note that I haven't put much work into docs here, that's left for #32 and #34

nagisa commented on src/lib.rs:
```diff
@@ -45,115 +46,300 @@ pub enum Protection {
+    pub fn make_mut(mut self, protection: Protection) -> Result<MmapMut> {
+        try!(self.inner.set_protection(protection));
+        match protection {
+            Protection::Read | Protection::ReadExecute => Err(Error::new(
```
IMO at this point it should just try to `set_protection` the mapping to the +w and only return Err if *that* fails.

bugaevc commented:
> Will take a look and likely merge tonight or over the weekend.

Any news?

danburkert commented:
Thanks!
